# Tutorial - Text Mining - Classification 

We will predict the category of discussion posts in a newsgroup.

**The unit of analysis is a discussion post**

### Import common packages

In [114]:
import pandas as pd
import numpy as np

np.random_seed = 1

### Load data

In [115]:
news = pd.read_csv('C:/Users/prath/Downloads/news.csv')

news.shape


(597, 5)

In [116]:
news.head(5)

TEXT  graphics  hockey  \
0  I have a few reprints left of chapters from my...         1       0   
1  gnuplot, etc. make it easy to plot real valued...         1       0   
2  Article-I.D.: snoopy.1pqlhnINN8k1 References: ...         1       0   
3  Hello, I am looking to add voice input capabil...         1       0   
4  I recently got a file describing a library of ...         1       0   

   medical newsgroup  
0        0  graphics  
1        0  graphics  
2        0  graphics  
3        0  graphics  
4        0  graphics

### Check for missing values

In [117]:
news[['TEXT']].isna().sum()

TEXT    0
dtype: int64

## Assign the input variable to X and the target variable to y

In [118]:
X = news['TEXT']

This is a multi-class classification problem. There are three categories we will predict:<br>
Whether a post is "graphics," "hockey," or "medical" related

In [119]:
y = news['newsgroup']
y.unique()

array(['graphics', 'hockey', 'medical'], dtype=object)

In [120]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()
le.fit(y)
print(le.classes_)
y = le.transform(y)

y


['graphics' 'hockey' 'medical']


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

## Split the data

In [121]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [122]:
X_train.shape, y_train.shape

((417,), (417,))

In [123]:
X_test.shape, y_test.shape

((180,), (180,))

In [124]:
X_train.head(5)

94     In article < jonas-y.734802983@gouraud> jonas-...
370    Article-I.D.: news.1993Apr15.181531.26088 Refe...
372    Article-I.D.: hydra.93158 References: < 93105....
496    In article < 1993Mar29.181406.11915@iscsvax.un...
30     WGT is the WordUp Graphics Toolkit, designed b...
Name: TEXT, dtype: object

In [125]:
y_train[:5]

array([0, 1, 1, 2, 0])

## Sklearn: Text preparation

For simplicity (and focus), we will not do any text cleaning or preprocessing. We will just use the raw text as input to the model. See the text mining fundamentals tutorial for more details on text cleaning and preprocessing.

In [126]:
#TfidfVectorizer includes pre-processing, tokenization, filtering stop words
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(stop_words='english', lowercase=True, token_pattern="[^\W\d_]+")

X_train = tfidf_vect.fit_transform(X_train)

**Notice in the previous step that we use `fit_transform` on TRAIN. When we transform the TEST data, we need to use `transform` only. This enables us to keep the number of columns (features) the same across the data sets. Otherwise, they WILL be different, and no model will work!**

In [127]:
# Perform the TfidfVectorizer transformation
# Be careful: We are using the train fit to transform the test data set. Otherwise, the test data 
# features will be very different and match the train set!!!

X_test = tfidf_vect.transform(X_test)


In [128]:
X_train.shape, X_test.shape

((417, 9989), (180, 9989))

In [129]:
# These data sets are "sparse matrix". We can't see them unless we convert using toarray()
X_train

<417x9989 sparse matrix of type '<class 'numpy.float64'>'
	with 30408 stored elements in Compressed Sparse Row format>

In [130]:
# These data sets are "sparse matrix". We can't see them unless we convert using toarray()
X_train.toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [86]:
performance = pd.DataFrame({"model": [],"SVD n_comp":[], "Accuracy": []})

## Latent Semantic Analysis (Singular Value Decomposition)

In [131]:
from sklearn.decomposition import TruncatedSVD

svd = TruncatedSVD(n_components=500, n_iter=10) #n_components is the number of topics, which should be less than the number of features

X_train= svd.fit_transform(X_train)
X_test = svd.transform(X_test)


In [132]:
X_train.shape, X_test.shape

((417, 417), (180, 417))

## Random Forest

In [133]:
from sklearn.ensemble import RandomForestClassifier 

rnd_clf = RandomForestClassifier(n_estimators=100, max_leaf_nodes=16, n_jobs=-1) 
_ = rnd_clf.fit(X_train, y_train)

### Evaluating Model Performance

In [134]:
from sklearn.metrics import accuracy_score

In [135]:
#Train accuracy - Not a good measure of model performance as we are using the same data set to train and test
y_pred_train = rnd_clf.predict(X_train)
acc = accuracy_score(y_train, y_pred_train)
print(f"Train acc: {accuracy_score(y_train, y_pred_train):.4f}")

Train acc: 0.9856


In [136]:
#Test accuracy
y_pred_test = rnd_clf.predict(X_test)
acc = accuracy_score(y_test, y_pred_test)
print(f"Test acc: {accuracy_score(y_test, y_pred_test)}")
performance = pd.concat([performance,pd.DataFrame({"model": 'Random Forest',"SVD n_comp":'500', "Accuracy": [acc]})])

Test acc: 0.8888888888888888


In [138]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred_test)

array([[47,  1, 12],
       [ 2, 56,  2],
       [ 2,  1, 57]], dtype=int64)

## Stochastic Gradient Descent Classifier

In [139]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(max_iter=100)
_ = sgd_clf.fit(X_train, y_train)

### Evaluating Model Performance

In [140]:
#Train accuracy
y_pred_train = sgd_clf.predict(X_train)
print(f"Train acc: {accuracy_score(y_train, y_pred_train):.4f}")

Train acc: 1.0000


In [141]:
#Test accuracy
y_pred_test = sgd_clf.predict(X_test)
acc1=accuracy_score(y_test, y_pred_test)
performance = pd.concat([performance,pd.DataFrame({"model": 'SGD Classifier',"SVD n_comp":'500', "Accuracy": [acc1]})])

In [142]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred_test)

array([[57,  0,  3],
       [ 1, 59,  0],
       [ 4,  2, 54]], dtype=int64)

In [147]:
performance.sort_values(by='model')

model SVD n_comp  Accuracy
0   Random Forest        300  0.861111
0   Random Forest        100  0.888889
0   Random Forest        500  0.888889
0  SGD Classifier        300  0.933333
0  SGD Classifier        100  0.955556
0  SGD Classifier        500  0.944444

In both the models, the accuracy is high when n_components=100. When n_components=500, the performance is slightly less in the case of SGD Classifier and in the case of Random Forest, it's the same as n_components=100. When n_components=300, the performance of both the models is less than the other two cases.

SVD is used to reduce the dimensions in other words extract the key features from large number of features. In our analysis it can be seen that there are large number of documents i.e., 9989 and in order for models to fit on the data it is important to provide them with key features than all of the features.  But there is a chance that it will discard the relevant features.